timelock-agent/src/main/java/com/palantir/timelock/config/KubernetesClusterConfiguration.java

```diff
@@ -0,0 +1,68 @@
+/*
+ * Copyright 2017 Palantir Technologies, Inc. All rights reserved.
+ *
+ * Licensed under the BSD-3 License (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://opensource.org/licenses/BSD-3-Clause
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.timelock.config;
+
+import java.util.List;
+import java.util.stream.Collectors;
+
+import org.immutables.value.Value;
+
+import com.fasterxml.jackson.annotation.JsonProperty;
+import com.fasterxml.jackson.databind.annotation.JsonDeserialize;
+import com.fasterxml.jackson.databind.annotation.JsonSerialize;
+import com.palantir.timelock.utils.KubernetesHostnames;
+
+/**
+ * Generates the current hostname, and those of the expected members of the cluster if the server is being executed
+ * within a Kubernetes stateful set.
+ */
+@Value.Immutable
+@JsonSerialize(as = ImmutableKubernetesClusterConfiguration.class)
+@JsonDeserialize(as = ImmutableKubernetesClusterConfiguration.class)
+public interface KubernetesClusterConfiguration extends ClusterConfiguration {
+
+    String TYPE = "kubernetes";
+
+    /** The number of members expected to join the cluster. */
+    @JsonProperty("expected-cluster-size")
+    int expectedClusterSize();
+
+    @JsonProperty("port")
+    int port();
+
+    /**
+     * The path to the api uri.
+     * <p>
+     * Used when generating the uris for the members of the cluster.
+     */
+    @JsonProperty("path")
+    String path();
+
+    @Override
+    default List<String> clusterMembers() {
+        return KubernetesHostnames.INSTANCE.getClusterMembers(expectedClusterSize())
+                .stream()
+                .map(hostname -> String.format("%s:%s%s", hostname, port(), path()))
+                .collect(Collectors.toList());
+    }
+
+    @Override
+    default String localServer() {
+        return String.format("%s:%s%s", KubernetesHostnames.INSTANCE.getCurrentHostname(), port(), path());
+    }
+
```

Comment:
We might want to add a `@Value.Check` here (that `expectedClusterSize` is at least 1, and maybe something about the port). To be fair some of this will be caught by the call to clusterMembers() in the default health-check, but might be better to have more specific "here's why your config doesn't work" messages.

Reply:
done